### PR TITLE
[REVIEW] Support for `cudf::table` to `arrow::Table` and `arrow::Table` to `cudf::table`

### LINKED ISSUE #1352: [FEA] To and from Arrow in CPP
**Is your feature request related to a problem? Please describe.**
Right now you can only go to and from arrow to cudf using python. We should be able to do this in cpp as well.

**Describe the solution you'd like**
To have to and from arrow methods in cpp.

**Describe alternatives you've considered**
None

**Additional context**
Users will often have arrow and those users might be consuming cudf using the cpp library. Particularly something as important as going to and from arrow.

### LINKED ISSUE #2210: [FEA] `to_arrow` should clear the padding bits in the bitmask
**Is your feature request related to a problem? Please describe.**
According to the Arrow spec:
> Bitmaps are to be initialized to be all unset at allocation time (this includes padding).

This would imply that bits outside the range `[0, size)` should always be zero. However, in cuDF/libcudf, we take a more conservative approach and say that bits outside `[0,size)` are undefined in order to avoid any potential issues.

The only potential issue I see is if an output from libcudf is consumed by some other library that supports Arrow, and that library makes assumptions about the state of the padding bits. In which case, we can guard against that in our `to_arrow` calls to zero out the padding bits.

**Describe the solution you'd like**
In `cuDF` and `libcudf` (which doesn't exist yet) `to_arrow` calls, the bits outside `[0, size)` should be cleared in accordance with Arrow spec.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Added test cases and all are working as expected apart from couple of blockers which inhibits couple test cases, list column creation from column_view is not yet supported.
 
closes #1352
closes #2210

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -417,6 +417,8 @@ add_library(cudf
             src/transform/transform.cpp
             src/transform/nans_to_nulls.cu
             src/transform/bools_to_mask.cu
+            src/transform/cudf_to_arrow.cpp
```
Comment (jrhemstad) [inline cpp/CMakeLists.txt:420]: These are not transform functions. They are for interop. We should make a new folder called `interop` and put DLPack + Arrow stuff in there.

File: cpp/src/transform/cudf_to_arrow.cpp [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+}  // namespace detail
+
+std::shared_ptr<arrow::Table> cudf_to_arrow(table_view input,
+                                            std::vector<std::string> const& column_names,
+                                            arrow::MemoryPool* ar_mr)
+{
+  auto num_columns = input.num_columns();
+
+  CUDF_EXPECTS((column_names.size() == num_columns),
```
Comment (jrhemstad) [inline cpp/src/transform/cudf_to_arrow.cpp:159]: The temporary is never used other than here. Just use it directly. 
```suggestion
  CUDF_EXPECTS((column_names.size() == input.num_columns()),
```

File: cpp/src/transform/cudf_to_arrow.cpp [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  }
+};
+
+}  // namespace detail
+
+std::shared_ptr<arrow::Table> cudf_to_arrow(table_view input,
+                                            std::vector<std::string> const& column_names,
+                                            arrow::MemoryPool* ar_mr)
+{
+  auto num_columns = input.num_columns();
```
Comment (jrhemstad) [inline cpp/src/transform/cudf_to_arrow.cpp:157]:
```suggestion
```

File: cpp/src/transform/cudf_to_arrow.cpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+  CUDF_EXPECTS((column_names.size() == num_columns),
+               "column names should be empty or should be equal to number of columns in table");
+
+  std::vector<std::shared_ptr<arrow::Array>> arrays;
+  std::vector<std::shared_ptr<arrow::Field>> fields;
+  std::shared_ptr<arrow::Schema> schema;
+  bool has_names = column_names.size() > 0;
```
Comment (jrhemstad) [inline cpp/src/transform/cudf_to_arrow.cpp:165]:
```suggestion
  bool const has_names = column_names.is_empty();
```

File: cpp/src/transform/cudf_to_arrow.cpp [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+  size_type itr = 0;
+  for (auto const& c : input) {
```
Comment (jrhemstad) [inline cpp/src/transform/cudf_to_arrow.cpp:168]: This should be two `transform`s.

File: cpp/src/transform/cudf_to_arrow.cpp [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+std::shared_ptr<arrow::Buffer> fetch_mask_buffer(column_view input_view, arrow::MemoryPool* ar_mr)
+{
+  const int64_t mask_size_in_bytes = cudf::bitmask_allocation_size_bytes(input_view.size());
+  std::shared_ptr<arrow::Buffer> mask_buffer;
+
+  if (input_view.has_nulls()) {
+    CUDF_EXPECTS(arrow::AllocateBuffer(ar_mr, mask_size_in_bytes, &mask_buffer).ok(),
+                 "Failed to allocate Arrow buffer for mask");
+    if (input_view.offset() > 0) {
+      cudaMemcpy(mask_buffer->mutable_data(),
```
Comment (jrhemstad) [inline cpp/src/transform/cudf_to_arrow.cpp:55]:
```suggestion
      CUDF_CUDA_TRY(cudaMemcpy(mask_buffer->mutable_data(),
```

File: cpp/src/transform/cudf_to_arrow.cpp [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    auto mask_buffer = fetch_mask_buffer(input_view, ar_mr);
+    std::vector<std::shared_ptr<arrow::Array>> child_arrays;
+
+    for (size_type i = 0; i < input_view.num_children(); i++) {
+      auto c = input_view.child(i);
+      child_arrays.emplace_back(
+        type_dispatcher(c.type(), dispatch_to_arrow{}, c, c.type().id(), ar_mr));
+    }
+
+    if (std::is_same<T, cudf::string_view>::value) {
```
Comment (jrhemstad) [inline cpp/src/transform/cudf_to_arrow.cpp:110]: Instead of putting all of the logic for all of the compound types in the same `is_compound` path, I think it would be much cleaner to have separate specializations for the different types.

File: cpp/src/transform/arrow_to_cudf.cu [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+                   std::pair<std::unique_ptr<column>, column_view>>
+  operator()(const std::shared_ptr<arrow::Array> arrow_col,
+             data_type type,
+             bool skip_mask,
+             rmm::mr::device_memory_resource* mr,
+             cudaStream_t stream)
+  {
+    CUDF_FAIL("Only fixed width and compund types are supported");
+  }
+};
```
Comment (jrhemstad) [inline cpp/src/transform/arrow_to_cudf.cu:225]: Anonymous namespace.

File: cpp/src/transform/arrow_to_cudf.cu [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+    return std::make_pair<std::unique_ptr<column>, column_view>(
+      std::move(col),
+      cudf::detail::slice(col->view(),
+                          static_cast<size_type>(arrow_col->offset()),
+                          static_cast<size_type>(arrow_col->length())));
+  }
+
+  template <typename T>
+  std::enable_if_t<is_compound<T>(), std::pair<std::unique_ptr<column>, column_view>> operator()(
```
Comment (jrhemstad) [inline cpp/src/transform/arrow_to_cudf.cu:93]: Same here. Use separate specializations.

File: cpp/src/transform/arrow_to_cudf.cu [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+    CUDF_FAIL("Only fixed width and compund types are supported");
+  }
+};
+
+std::unique_ptr<table> arrow_to_cudf(std::shared_ptr<arrow::Table> input_table,
+                                     rmm::mr::device_memory_resource* mr,
+                                     cudaStream_t stream)
+{
+  std::vector<std::unique_ptr<column>> columns;
+  for (auto const& chunks_col : input_table->columns()) {
```
Comment (jrhemstad) [inline cpp/src/transform/arrow_to_cudf.cu:232]: This is a `transform`.

File: cpp/src/transform/arrow_to_cudf.cu [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+    case arrow::Type::type::DICTIONARY: return data_type(type_id::DICTIONARY32);
+    case arrow::Type::type::LIST: return data_type(type_id::LIST);
+    default: CUDF_FAIL("Unsupported type_id conversion to cudf");
+  }
+}
+
+struct dispatch_to_cudf_col {
+  template <typename T>
+  std::enable_if_t<is_fixed_width<T>(), std::pair<std::unique_ptr<column>, column_view>> operator()(
+    const std::shared_ptr<arrow::Array> arrow_col,
```
Comment (jrhemstad) [inline cpp/src/transform/arrow_to_cudf.cu:62]: Don't perpetuate Arrow's bad style of unnecessarily passing by `shared_ptr`. 
```suggestion
    arrow::Array const& arrow_col,
```

File: cpp/src/transform/arrow_to_cudf.cu [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+    return std::make_pair<std::unique_ptr<column>, column_view>(
+      std::move(col),
+      cudf::detail::slice(col->view(),
+                          static_cast<size_type>(arrow_col->offset()),
+                          static_cast<size_type>(arrow_col->length())));
+  }
+
+  template <typename T>
+  std::enable_if_t<is_compound<T>(), std::pair<std::unique_ptr<column>, column_view>> operator()(
+    const std::shared_ptr<arrow::Array> arrow_col,
```
Comment (jrhemstad) [inline cpp/src/transform/arrow_to_cudf.cu:94]:
```suggestion
    arrow::Array const& arrow_col,
```

File: cpp/include/cudf/transform.hpp [hunk 13]
```diff
@@ -83,6 +86,36 @@ std::pair<std::unique_ptr<rmm::device_buffer>, size_type> nans_to_nulls(
...
+ * metadata `column_names`.
+ *
+ * @throws cudf::logic_error if `column_names` size doesn't match with number of columns.
+ *
+ * @param input table_view that needs to be converted to arrow Table
+ * @param column_names Vector of column names for metadata of arrow Table
+ * @param ar_mr arrow memory pool to allocate memory for arrow Table
+ * @return arrow Table generated from given input cudf table
+ **/
+std::shared_ptr<arrow::Table> cudf_to_arrow(
```
Comment (jrhemstad) [inline cpp/include/cudf/transform.hpp:102]:
```suggestion
std::unique_ptr<arrow::Table> cudf_to_arrow(
```

File: cpp/include/cudf/transform.hpp [hunk 14]
```diff
@@ -83,6 +86,36 @@ std::pair<std::unique_ptr<rmm::device_buffer>, size_type> nans_to_nulls(
...
+/**
+ * @brief Create `cudf::table` from given arrow Table input
+ *
+ * @param input  arrow:Table that needs to be converted to `cudf::table`
+ * @param mr           Device memory resource used to allocate `cudf::table`
+ * @return cudf table generated from given arrow Table.
+ **/
+
+std::unique_ptr<table> arrow_to_cudf(
+  std::shared_ptr<arrow::Table> input_table,
```
Comment (jrhemstad) [inline cpp/include/cudf/transform.hpp:116]: We shouldn't perpetuate Arrow's style of unnecessarily passing by `shared_ptr`. This function doesn't take ownership or extend the lifetime of the `input_table`.  
```suggestion
  arrow::Table const& input_table,
```

**Comment by jrhemstad** [comment]
Oh, and this PR should take care of this as well: https://github.com/rapidsai/cudf/issues/2210

File: cpp/src/transform/arrow_to_cudf.cu (hunk 12 above)
Comment (rgsl888prabhu) [reply cpp/src/transform/arrow_to_cudf.cu:94]: I don't think this will help as under compound types, I need to extract child class and making Array will omit subclass information.

File: cpp/include/cudf/transform.hpp (hunk 14 above)
Comment (rgsl888prabhu) [reply cpp/include/cudf/transform.hpp:116]: And even this can't be done as l Table COPY and ASSIGN constructors are private.

File: cpp/include/cudf/transform.hpp (hunk 13 above)
Comment (rgsl888prabhu) [reply cpp/include/cudf/transform.hpp:102]: arrow::Table can be created with only arrow::Table::Make which returns a shared pointer, and moving a shared pointer to a unique pointer doesn't feel right to me.

File: cpp/src/transform/arrow_to_cudf.cu (hunk 12 above)
Comment (jrhemstad) [reply cpp/src/transform/arrow_to_cudf.cu:94]: It's still being passed by reference, so you can still use it polymorphically.

File: cpp/include/cudf/transform.hpp (hunk 13 above)
Comment (jrhemstad) [reply cpp/include/cudf/transform.hpp:102]: Ah, well then that's Arrow's fault for returning `shared_ptr` by default. I guess there's nothing we can do about it.

File: cpp/include/cudf/transform.hpp (hunk 14 above)
Comment (jrhemstad) [reply cpp/include/cudf/transform.hpp:116]: I don't understand. We're passing by `const` reference. There's no need for copy ctors.

File: cpp/src/transform/arrow_to_cudf.cu (hunk 12 above)
Comment (rgsl888prabhu) [reply cpp/src/transform/arrow_to_cudf.cu:94]: We get shared_ptr<arrow::Array> from arrow::Table, so dreferencing will give us just arrow::Array, and we will not be able to cast it to anything.

Comment (jrhemstad) [reply cpp/src/transform/arrow_to_cudf.cu:94]: I don't understand. In terms of polymorphism, there is nothing a `shared_ptr` can do that a `const&` cannot do.

File: cpp/src/transform/arrow_to_cudf.cu [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+      auto char_array = std::make_shared<arrow::Int8Array>(
+        str_array->value_data()->size(), str_array->value_data(), nullptr);
+
+      auto offsets_column = std::move(type_dispatcher(data_type(type_id::INT32),
+                                                      dispatch_to_cudf_col{},
+                                                      offset_array,
+                                                      data_type(type_id::INT32),
+                                                      true,
+                                                      mr,
+                                                      stream)
```
Comment (davidwendt) [inline cpp/src/transform/arrow_to_cudf.cu:125]: Seems a waste to call the `type_dispatcher` for a hardcoded type. Can the functor just be called directly?
```
auto offsets_column = std::move(
dispatch_to_cudf_col{}.operator()<int32_t>(offset_array,
                                           data_type(type_id::INT32), true, mr, stream)
.first);
```

File: cpp/src/transform/arrow_to_cudf.cu [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+    bool skip_mask,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    std::unique_ptr<rmm::device_buffer> mask = std::make_unique<rmm::device_buffer>(0);
+    std::unique_ptr<column> out_col;
+
+    if (arrow_col->null_bitmap_data() != nullptr) {
+      mask = std::make_unique<rmm::device_buffer>(arrow_col->null_bitmap()->size(), stream, mr);
+      cudaMemcpy(mask->data(),
```
Comment (davidwendt) [inline cpp/src/transform/arrow_to_cudf.cu:105]: cudaMemcpyAsync here?

File: cpp/src/transform/arrow_to_cudf.cu (hunk 15 above)
Comment (rgsl888prabhu) [reply cpp/src/transform/arrow_to_cudf.cu:125]: That's a nice option.

File: cpp/src/interop/cudf_to_arrow.cpp [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+    std::vector<std::shared_ptr<arrow::Array>> child_arrays;
+    std::vector<size_type> child_indices(input_view.num_children());
+    std::iota(child_indices.begin(), child_indices.end(), 0);
+    std::transform(child_indices.begin(),
+                   child_indices.end(),
+                   std::back_inserter(child_arrays),
+                   [&input_view, &ar_mr, &stream](auto const& i) {
+                     auto c = input_view.child(i);
+                     return type_dispatcher(
+                       c.type(), dispatch_to_arrow{}, c, c.type().id(), ar_mr, stream);
```
Comment (davidwendt) [inline cpp/src/interop/cudf_to_arrow.cpp:92]: This does not appear to handle the `input_view.offset()` value.

File: cpp/include/cudf/detail/interop.hpp [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+
+// Converting arrow type to cudf type
+data_type arrow_to_cudf_type(arrow::Type::type arrow_type);
+
+/**
+ * @copydoc cudf::cudf_to_arrow
+ *
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ **/
+std::shared_ptr<arrow::Table> cudf_to_arrow(table_view input,
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/interop.hpp:78]:
```suggestion
std::shared_ptr<arrow::Table> to_arrow(table_view input,
```

File: cpp/include/cudf/detail/interop.hpp [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+                                            std::vector<std::string> const& column_names = {},
+                                            arrow::MemoryPool* ar_mr = arrow::default_memory_pool(),
+                                            cudaStream_t stream      = 0);
+
+/**
+ * @copydoc cudf::arrow_to_cudf
+ *
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ **/
+std::unique_ptr<table> arrow_to_cudf(
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/interop.hpp:88]:
```suggestion
std::unique_ptr<table> from_arrow(
```

File: cpp/include/cudf/interop.hpp [hunk 20]
```diff
@@ -68,5 +73,41 @@ std::unique_ptr<table> from_dlpack(
 DLManagedTensor* to_dlpack(table_view const& input,
                            rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/** @} */  // end of group
+
+/**
+ * @addtogroup interop_arrow
```
Comment (davidwendt) [inline cpp/include/cudf/interop.hpp:79]: Add this new group to `cpp/include/doxygen_groups.h`
https://github.com/rapidsai/cudf/blob/8fe2b9214a8cbed70cc4b6ee169e8d3ae201b76f/cpp/include/doxygen_groups.h#L86-L89

File: cpp/src/interop/arrow_to_cudf.cu [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+    if (array.null_bitmap_data() != nullptr) {
+      mask = std::make_unique<rmm::device_buffer>(array.null_bitmap()->size(), stream, mr);
+      CUDA_TRY(cudaMemcpyAsync(mask->data(),
+                               array.null_bitmap_data(),
+                               array.null_bitmap()->size(),
+                               cudaMemcpyHostToDevice,
+                               stream));
+    }
+
+    return std::move(mask);
```
Comment (jrhemstad) [inline cpp/src/interop/arrow_to_cudf.cu:107]:
```suggestion
    return mask;
```

File: cpp/src/interop/arrow_to_cudf.cu [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+                          static_cast<size_type>(array.offset()),
+                          static_cast<size_type>(array.offset() + array.length())));
+  }
+
+  std::unique_ptr<rmm::device_buffer> get_mask_buffer(arrow::Array const& array,
+                                                      rmm::mr::device_memory_resource* mr,
+                                                      cudaStream_t stream)
+  {
+    std::unique_ptr<rmm::device_buffer> mask = std::make_unique<rmm::device_buffer>(0);
+    if (array.null_bitmap_data() != nullptr) {
```
Comment (jrhemstad) [inline cpp/src/interop/arrow_to_cudf.cu:98]: ITM. Use an immediately invoke lambda instead.

File: cpp/src/interop/arrow_to_cudf.cu [hunk 23]
```diff
@@ -94,17 +94,20 @@ struct dispatch_to_cudf_column {
...
-                               cudaMemcpyHostToDevice,
-                               stream));
-    }
-
-    return std::move(mask);
+    return [&]() {
+      if (array.null_bitmap_data() == nullptr) {
+        return std::make_unique<rmm::device_buffer>(0);
+      } else {
+        std::unique_ptr<rmm::device_buffer> mask = std::make_unique<rmm::device_buffer>(0);
```
Comment (jrhemstad) [inline cpp/src/interop/arrow_to_cudf.cu:101]:
```suggestion
```

File: cpp/src/interop/arrow_to_cudf.cu [hunk 24]
```diff
@@ -94,17 +94,20 @@ struct dispatch_to_cudf_column {
...
+        mask = std::make_unique<rmm::device_buffer>(array.null_bitmap()->size(), stream, mr);
```
Comment (jrhemstad) [inline cpp/src/interop/arrow_to_cudf.cu:102]:
```suggestion
        auto mask = std::make_unique<rmm::device_buffer>(array.null_bitmap()->size(), stream, mr);
```

File: cpp/src/interop/arrow_to_cudf.cu [hunk 25]
```diff
@@ -94,17 +94,20 @@ struct dispatch_to_cudf_column {
...
-                               array.null_bitmap_data(),
-                               array.null_bitmap()->size(),
-                               cudaMemcpyHostToDevice,
-                               stream));
-    }
-
-    return std::move(mask);
+    return [&]() {
+      if (array.null_bitmap_data() == nullptr) {
+        return std::make_unique<rmm::device_buffer>(0);
```
Comment (jrhemstad) [inline cpp/src/interop/arrow_to_cudf.cu:99]:
```suggestion
        return std::make_unique<rmm::device_buffer>(0, stream, mr);
```

File: cpp/src/interop/arrow_to_cudf.cu [hunk 26]
```diff
@@ -0,0 +1,297 @@
...
+      } else {
+        auto mask = std::make_unique<rmm::device_buffer>(array.null_bitmap()->size(), stream, mr);
+        CUDA_TRY(cudaMemcpyAsync(mask->data(),
+                                 array.null_bitmap_data(),
+                                 array.null_bitmap()->size(),
+                                 cudaMemcpyHostToDevice,
+                                 stream));
+        return mask;
+      }
+    }();
```
Comment (trxcllnt) [inline cpp/src/interop/arrow_to_cudf.cu:109]: Is there a reason this is wrapped in an immediately-executed function expression?

Comment (rgsl888prabhu) [reply cpp/src/interop/arrow_to_cudf.cu:109]: Just review comment to improvise.

Comment (rgsl888prabhu) [reply cpp/src/interop/arrow_to_cudf.cu:109]: https://github.com/rapidsai/cudf/pull/5778#discussion_r462562920

Comment (harrism) [reply cpp/src/interop/arrow_to_cudf.cu:109]: There's a "Better Code" recording on this anti-pattern. :)

Comment (rgsl888prabhu) [reply cpp/src/interop/arrow_to_cudf.cu:109]: I think this is the pattern which was discussed to be better than if else cond, if I am not wrong.

Comment (davidwendt) [reply cpp/src/interop/arrow_to_cudf.cu:109]: Please, just no.

Comment (trxcllnt) [reply cpp/src/interop/arrow_to_cudf.cu:109]: @rgsl888prabhu if I understand correctly, the original code implemented the "define then modify" anti-pattern with the null mask:
```c++
auto mask = rmm:device_buffer(0);
if (array.null_bitmap_data()) { mask = rmm::device_buffer(array.null_bitmap()); }
return std::make_unique(mask);
```

But the current version with two return statements avoids that anti-pattern by returning early, and the IIFE isn't adding anything. Am I misunderstanding something fundamental that the lambda is enabling?

Comment (rgsl888prabhu) [reply cpp/src/interop/arrow_to_cudf.cu:109]: No, I think its just same nothing different.

File: cpp/src/interop/arrow_to_cudf.cu [hunk 27]
```diff
@@ -0,0 +1,302 @@
+/*
```
Comment (jrhemstad) [inline cpp/src/interop/arrow_to_cudf.cu:1]: This file should be renamed to `from_arrow.cu`.

File: cpp/src/interop/cudf_to_arrow.cpp [hunk 28]
```diff
@@ -0,0 +1,227 @@
+/*
```
Comment (jrhemstad) [inline cpp/src/interop/cudf_to_arrow.cpp:1]: `to_arrow.cpp`

File: cpp/src/interop/arrow_to_cudf.cu [hunk 29]
```diff
@@ -0,0 +1,302 @@
...
+    return mask;
+  }
+};
+
+/**
+ * @brief Returns cudf column formed from given arrow array
+ * This has been introduced to take care of compile error "error: explicit specialization of
+ * function must precede its first use"
+ */
+std::pair<std::unique_ptr<column>, column_view> get_column(arrow::Array const& array,
```
Comment (jrhemstad) [inline cpp/src/interop/arrow_to_cudf.cu:120]: Why does this return a `column` and a `column_view`?

File: cpp/src/interop/arrow_to_cudf.cu [hunk 30]
```diff
@@ -0,0 +1,302 @@
...
+                               array.null_bitmap()->size(),
+                               cudaMemcpyHostToDevice,
+                               stream));
+    }
+
+    return std::make_pair<std::unique_ptr<column>, column_view>(
+      std::move(col),
+      cudf::detail::slice(col->view(),
+                          static_cast<size_type>(array.offset()),
+                          static_cast<size_type>(array.offset() + array.length())));
```
Comment (jrhemstad) [inline cpp/src/interop/arrow_to_cudf.cu:95]: I see, this function returns a `column` and `column_view` because it copies the entire `arrow::Array` data buffer even if it has an offset. This seems wasteful.

Why not only copy the data in the `arrow::Array` from `[offset, size)`?

File: cpp/src/interop/arrow_to_cudf.cu [hunk 31]
```diff
@@ -0,0 +1,302 @@
...
+                       concat_column_views.emplace_back(col_and_view.second);
+                       return std::move(col_and_view.first);
+                     });
+
+                   if (concat_columns.size() == 1) {
+                     if ((concat_column_views[0].offset() == 0) and
+                         (concat_column_views[0].size() == concat_columns[0]->size())) {
+                       return std::move(concat_columns[0]);
+                     } else {
+                       return std::make_unique<column>(concat_column_views[0]);
```
Comment (jrhemstad) [inline cpp/src/interop/arrow_to_cudf.cu:283]: This is making another copy just to account for the `offset`! 

See my above comment to make `get_column` only copy `[offset, size)` from the `arrow::Array`.

File: cpp/src/interop/arrow_to_cudf.cu (hunk 30 above)
Comment (rgsl888prabhu) [reply cpp/src/interop/arrow_to_cudf.cu:95]: In case of string or list, the offsetting the child column doesn't make any sense. And for which you have to completely form the column and then take the offset.

Comment (rgsl888prabhu) [reply cpp/src/interop/arrow_to_cudf.cu:95]: Else I can  handle fixed width completely different and child columns of compound times completely different.

Comment (jrhemstad) [reply cpp/src/interop/arrow_to_cudf.cu:95]: I'm not following what the problem is.

Comment (rgsl888prabhu) [reply cpp/src/interop/arrow_to_cudf.cu:95]: I had started with some approach with offsets in mind, but with end changes I think it may not be required. Will just copy from offset to size as you mentioned.

Comment (trxcllnt) [reply cpp/src/interop/arrow_to_cudf.cu]: I'm not sure if this is handled, but in case not -- we need to be sure to rebase the codes from the offset position when copying sliced String, and List columns to/from Arrow:
```
// original Arrow::StringArray or cudf::column
string_col = { offset: 0, offsets: [0,1,2,3], data: ["abcd"] }

// sliced Arrow::StringArray or cudf::column (offsets sliced, data untouched)
offset_col = { offset: 2, offsets: [2,3], data: ["abcd"] }

// result of to_arrow or from_arrow with `offset_col` input
result_col = { offset: 0, offsets: [0,1], data: ["cd"] }
```

Comment (rgsl888prabhu) [reply cpp/src/interop/arrow_to_cudf.cu]: I am copying whole offsets and data and then slicing it in cudf, https://github.com/rapidsai/cudf/blob/0794298c5dbd2a01fa0130aa10a7565f3dbffaea/cpp/src/interop/from_arrow.cu#L173

File: cpp/include/cudf/transform.hpp [hunk 32]
```diff
@@ -111,5 +111,30 @@ std::pair<std::unique_ptr<cudf::column>, std::unique_ptr<cudf::column>> encode(
...
+ * output: [{false, true, false, true, false, true, false, true}]
+ * @endcode
+ *
+ * @param bitmask A device pointer to the bitmask which needs to be converted
+ * @param offset position of the bit from which the conversion should start
+ * @param length Number of bits needs to be converted from offset position
+ * @param mr Device memory resource used to allocate the returned columns's device memory
+ * @return A boolean column representing the given mask from [offset, offset+length).
+ */
+std::unique_ptr<column> mask_to_bools(
```
Comment (jrhemstad) [inline cpp/include/cudf/transform.hpp:133]: This should be a separate PR.

Comment (rgsl888prabhu) [reply cpp/include/cudf/transform.hpp:133]: It is quite small, so thought I could squeeze it in :)

File: cpp/src/interop/to_arrow.cpp [hunk 33]
```diff
@@ -0,0 +1,271 @@
...
+                        data_buffer,
+                        fetch_mask_buffer(input, ar_mr, stream),
+                        static_cast<int64_t>(input.null_count()));
+}
+
+template <>
+std::shared_ptr<arrow::Array> dispatch_to_arrow::operator()<cudf::string_view>(
+  column_view input, cudf::type_id id, arrow::MemoryPool* ar_mr, cudaStream_t stream)
+{
+  std::unique_ptr<column> tmp_column = nullptr;
```
Comment (jrhemstad) [inline cpp/src/interop/to_arrow.cpp:152]: ITM.
